### PR TITLE
Footnotes: disable for synced patterns and prevent duplication for pages in site editor

### PR DESCRIPTION
**Pull request by ramonjd**
## What?
A follow up to:

- https://github.com/WordPress/gutenberg/pull/52934

This PR attempts to resolves [some issues](https://github.com/WordPress/gutenberg/pull/52934#pullrequestreview-1546659253) found while testing:

- Prevents footnote creation within `core/block` blocks
- Only insert a footnote if one isn't found in the entity block list

## Why?
See https://github.com/WordPress/gutenberg/pull/52934#pullrequestreview-1546659253

## How?

- Prevent footnote creation within core/block
- Only insert a footnote if one isn't found in the entity block list

## Testing Instructions

- Create a new synced pattern. You shouldn't be able to add footnotes when editing the pattern in the site and post editors.
- Double check this for unsynced patterns
- In the site editor, edit a page that already has footnotes, and try to add some more. Ensure that the footnote block is not duplicated

### CODE REVIEW

File: packages/block-library/src/footnotes/format.js
Comment (ellatrix) [inline packages/block-library/src/footnotes/format.js]: `blocks` is only used in the onClick callback. Is there a selector that we could run there instead?

Comment (ramonjd) [inline packages/block-library/src/footnotes/format.js]: Added this in the first version only to kick things off.

@ellatrix identified [potential performance implications](https://github.com/WordPress/gutenberg/pull/52934#issuecomment-1651270665):

> useEntityBlockEditor would mean it's re-rendering for any other block changes, it would be best to call a selector in the event handler.

Also I noticed that the footnotes linger when I create a synced pattern from blocks that already have footnotes.

I believe this is also an existing issue, but just making a note here that maybe we could warn folks that their footnotes will no longer work if they create a pattern, and then strip the footnotes? Not sure what the best UX would be.

Comment (ramonjd) [reply packages/block-library/src/footnotes/format.js]: Oh I was just writing the [comment](https://github.com/WordPress/gutenberg/pull/53003/files#r1275542391) below!

Looking into it now. Thanks!

Comment (ramonjd) [reply packages/block-library/src/footnotes/format.js]: I need to look further, but I'm having trouble finding an alternative that gives me the rendered block list for a page in the site editor. It's definitely a limitation of my knowledge, I don't spend much time in core-data 😄 

The `blocks` property is available outside the click handler, but not within it for example

```js
// returns the right block list outside onClick but `undefined` within it
getEditedEntityRecord(
					'postType',
					postType,
					postId
				)?.blocks
```

I can see that we get the edited entity record with raw content, whose blocks could be parsed. Not sure if that's very performant either, of if we'd get the necessary clientId of the footnotes block.

File: packages/block-library/src/footnotes/format.js
Comment (ramonjd) [inline packages/block-library/src/footnotes/format.js]: This is interesting.

Pursuant to the [comment above](https://github.com/WordPress/gutenberg/pull/53003/files#r1275543962), `getEditedEntityRecord` _kinda_ works:

```js
				const blocks = getEditedEntityRecord(
					'postType',
					postType,
					postId
				)?.blocks;

				// Finds the first footnote block.
				let fnBlock = blocks?.find( ( block ) => block.name === name );
```

But first we have to interact with the page itself, e.g., by editing the content in any way



https://github.com/WordPress/gutenberg/assets/6458278/56b4f56e-aeab-4784-9832-4cbb1741c945


`postType` and  `postId` are available in the callback on first render, so it isn't that I believe

But we're getting closer 😄 

Is there a way to trigger this interactivity programmatically? 🤷🏻

File: packages/block-library/src/footnotes/format.js
Comment (andrewserong) [reply packages/block-library/src/footnotes/format.js]: > I can see that we get the edited entity record with raw content, whose blocks could be parsed. Not sure if that's very performant either

I think using `getEditedEntityRecord` gets us closer to the desired end state. From looking at `useEntityBlockEditor`, it currently does the parsing [here](https://github.com/WordPress/gutenberg/blob/dff466a315104ba36585f571f1d836643f789d95/packages/core-data/src/entity-provider.js#L180) if `blocks` isn't available, so performance-wise, if we copy/pasted (or followed) a similar approach in the `onClick` callback, it'd be no slower than using `useEntityBlockEditor`, I'd think.

> if we'd get the necessary clientId of the footnotes block.

It doesn't look like we need the clientId of the footnotes block, rather, we're just checking whether it exists. So I _think_ we should be fine using the parsed content 🤞

Comment (andrewserong) [reply packages/block-library/src/footnotes/format.js]: > if we'd get the necessary clientId of the footnotes block.

Oh, wait, that does seem to be problem unfortunately! Sorry I missed that in my first pass, without that clientId, we don't have selection move down to the footnote block 🤔

File: packages/block-library/src/footnotes/format.js
Comment (tellthemachines) [reply packages/block-library/src/footnotes/format.js]: I can't say I'm hugely knowledgeable about the core-data stuff either, but would `getEntityRecord` work for retrieving the initial state?

Comment (andrewserong) [reply packages/block-library/src/footnotes/format.js]: I think part of the problem might be that in order for the `selectionChange()` call below to work, we might need to somehow get the same list of blocks (and therefore clientIds) that gets set within the parent post content block's `useInnerBlocks` [here](https://github.com/WordPress/gutenberg/blob/fbc67a11105e5884c05a8498b01c70173ad181ee/packages/block-library/src/post-content/edit.js#L70) (the post content block's controlled inner blocks).

From playing around locally, when I've attempted to call `getEditedEntityRecord` or `getEntityRecord` to grab the data I need, we wind up with a duplicate of the blocks, rather than a reference to the existing Footnotes block in the editor 🤔

**Comment by andrewserong** [comment]
Update: Hope you don't mind @ramonjd, but I've pushed a change: I've had a go at attempting to grab the controlled inner blocks of the parent post content block in https://github.com/WordPress/gutenberg/pull/53003/commits/f6087c7dd1f59a86744a13cc19461159bb6e8001 — I think this means we can avoid having to call `getEditedEntityRecord` directly, however in testing it still doesn't appear to be redirecting focus correctly to the found footnotes block within the page editor within the site editor, so it might not be quite right yet.

Do feel free to revert if this approach is no good, of course!

**Comment by ramonjd** [comment]
> Hope you don't mind @ramonjd, but I've pushed a change

Not at all. Thanks for stepping in here. I'll try to come back to it tomorrow (secretly hoping it'll be "magically" fixed overnight :trollface:)

**Comment by ramonjd** [comment]
> still doesn't appear to be redirecting focus correctly to the found footnotes block within the page editor within the site editor, so it might not be quite right yet.

That was happening in trunk as well, so I think we're good. 

Thanks for looking at this while I was out. Using `getBlocks( parentPostContent[ 0 ] )` seems to address the main functional concerns it seems from manual testing. At least I think it's an improvement. 😄 

Let's see what @ellatrix and others think.

File: packages/block-library/src/footnotes/format.js
Comment (ellatrix) [inline packages/block-library/src/footnotes/format.js]: Minor: Is there any way to check for this without hardcoding the block name?

File: packages/block-library/src/footnotes/format.js
Comment (ellatrix) [inline packages/block-library/src/footnotes/format.js]: We shouldn't be calling selectors like that in a component render call. This is only fine in event callbacks. We should create a separate `useSelect` call that listens for changes to the store so the component re-renders when the result of this function changes.

File: packages/block-library/src/footnotes/format.js
Comment (ellatrix) [inline packages/block-library/src/footnotes/format.js:84]: Could we not simply check for the presence of a post id? Would there be a post id in context when you select inside a template part?

Comment (ellatrix) [reply packages/block-library/src/footnotes/format.js:84]: Oh, ignore me, I thought this was for template parts.

File: packages/block-library/src/footnotes/format.js
Comment (ellatrix) [inline packages/block-library/src/footnotes/format.js:121]: Nit: we could be using the found root client id below when inserting a footnotes if none is present.

File: packages/block-library/src/footnotes/format.js
Comment (ramonjd) [reply packages/block-library/src/footnotes/format.js]: Good question. 

I'm not sure about that, but maybe @glendaviesnz or @aaronrobertshaw might have a better idea about to detect whether a block is inside a pattern? 

Context for those folks: here we want to know to know if any of a block's parents are  patterns by filtering parents using `core/block`.

Comment (ramonjd) [reply packages/block-library/src/footnotes/format.js]: Oh I just discovered `isReusableBlock`

Comment (ramonjd) [reply packages/block-library/src/footnotes/format.js]: > Oh I just discovered isReusableBlock 

Doesn't help here 😄

File: packages/block-library/src/footnotes/format.js
Comment (ramonjd) [reply packages/block-library/src/footnotes/format.js]: Oh good one, thanks for the nudge.

I've done this, slightly verbosely... open to optimization recommendations 😄

File: packages/block-library/src/footnotes/format.js
Comment (aaronrobertshaw) [reply packages/block-library/src/footnotes/format.js]: Also, `isReusableBlock` hard codes the `core/block` name as well. 

No alternatives spring to mind but @glendaviesnz has explored this area more so might have better ideas.

Comment (glendaviesnz) [reply packages/block-library/src/footnotes/format.js]: To-date I have only used `isReusableBlock` sorry, I don't think there is currently an alternative but maybe it is worth adding a new selector along the lines of `isParentSyncedPattern`?

Comment (ramonjd) [reply packages/block-library/src/footnotes/format.js]: >  I don't think there is currently an alternative but maybe it is worth adding a new selector along the lines of isParentSyncedPattern?

Sounds like a useful follow up. Thank you!

File: packages/block-library/src/footnotes/format.js
Comment (andrewserong) [inline packages/block-library/src/footnotes/format.js]: Nit: `hasParentCodeBlocks` could sound like it means whether or not there is any core block as a parent. Would something like `isBlockWithinPattern` be more direct?

Comment (andrewserong) [inline packages/block-library/src/footnotes/format.js]: Another tiny nit (not a blocker), would it be worth adding a comment before this `useSelect` to explain the decision to have a separate `useSelect`? I could imagine someone else in the future looking at this and thinking they could optimise it by re-using the `getSelectedBlockClientId` and `getBlockParentsByBlockName` functions above, and accidentally adversely affect performance in the process.

**Review by andrewserong — APPROVED** [review]
> > still doesn't appear to be redirecting focus correctly to the found footnotes block within the page editor within the site editor, so it might not be quite right yet.
>
> That was happening in trunk as well, so I think we're good.

Agreed 👍

This is working nicely for me and the code looks neat. You might want to get a final review from @ellatrix before landing, but this covers things nicely for now IMO:

✅ Footnotes cannot be added in synced patterns edited within the post editor
✅ Footnotes cannot be added in synced patterns within the site editor
✅ For pages edited within the site editor, duplicate Footnotes blocks are no longer appearing
✅ Otherwise editing footnotes appears to be unaffected

Just left a couple of tiny non-blocking nits / questions, but nothing urgent.

LGTM! ✨

File: packages/block-library/src/footnotes/format.js
Comment (ramonjd) [reply packages/block-library/src/footnotes/format.js]: Fair point. Will change quickly. Thanks for re-testing this (and for committing the improvement)

File: packages/block-library/src/footnotes/format.js
Comment (ellatrix) [inline packages/block-library/src/footnotes/format.js:154]: Nit: Why do we still need a while loop here if we got the parent post content block earlier?

Comment (ellatrix) [reply packages/block-library/src/footnotes/format.js:154]: In other words, if we know if there's a post content block, we can append it there, if there's no such block, append it to `''` (the root)?

**Review by ellatrix — APPROVED** [review]
Works great!

File: packages/block-library/src/footnotes/format.js
Comment (ramonjd) [reply packages/block-library/src/footnotes/format.js:154]: Oh, good point. 

I'm testing replacing with 

```js
					if ( parentPostContent.length ) {
						rootClientId = parentPostContent[ 0 ];
					}
```

Working okay so far 👍🏻

File: packages/block-library/src/footnotes/format.js
Comment (ramonjd) [inline packages/block-library/src/footnotes/format.js:151]: I think the failing test in footnotes.spec.js is related to this change? I might have to revert and do a follow up if so.

File: packages/block-library/src/footnotes/format.js
Comment (ramonjd) [inline packages/block-library/src/footnotes/format.js]: I think the failing test in footnotes.spec.js is related to this change? (no longer using `while`) 

I might have to revert and do a follow up if so.

Comment (ramonjd) [reply packages/block-library/src/footnotes/format.js]: yep, going to have to revert. I think it's related to nested blocks. 
Maybe we can tidy up in a follow up.

File: packages/block-library/src/footnotes/format.js
Comment (ramonjd) [reply packages/block-library/src/footnotes/format.js:154]: I had to revert this change because adding footnotes to list blocks was failing. I suspect this test doesn't account for nested blocks. Given we're close to 6.3 release,  we can do a follow up to see if we can tidy this.

Comment (ramonjd) [reply packages/block-library/src/footnotes/format.js:154]: The failing test that brought this to our attention was `Footnotes › can be inserted in a list` for the record. Great to have tests!!

**Comment by tellthemachines** [comment]
I just cherry-picked this PR to the update/second-round-6-3-rc3 branch to get it included in the next release: ef4f1f2cbc

**Pull request merged**
